homeassistant/components/homekit/util.py

```diff
@@ -110,6 +113,59 @@ def validate_media_player_features(state, feature_list):
     return True
 
 
+SpeedRange = namedtuple('SpeedRange', 'start target stop')
+SpeedRange.__doc__ += """: a mapping of descrete Home Assistant speed \
+values to percentage based HomeKit speeds."""
+SpeedRange.start.__doc__ += """Start of the range (inclusive)."""
+SpeedRange.target.__doc__ += """Percentage to use to determine HomeKit \
+percentages from HomeAssistant speed."""
+SpeedRange.stop.__doc__ += """Stop of the range (exclusive)."""
+
+
+class HomeKitSpeedMapping:
+    """Supports conversion between Home Assistant and HomeKit fan speeds."""
+
+    def __init__(self, speed_list):
+        """Initialize a new SpeedMapping object."""
+        if not speed_list or speed_list[0] != fan.SPEED_OFF:
+            _LOGGER.warning("%s does not contain the speed setting "
+                            "%s as its first element. "
+                            "Assuming that %s is equivalent to 'off'.",
+                            speed_list, fan.SPEED_OFF, speed_list[0])
+        speed_ranges = {}
+        list_size = len(speed_list)
+        for index, speed in enumerate(speed_list):
+            # By dividing by list_size -1 the following
+            # desired attributes hold true:
+            # * index = 0 => 0%, equal to "off"
+            # * index = len(speed_list) - 1 => 100 %
+            # * all other indices are equally distributed
+            target = index * 100 / (list_size - 1)
+            start = index * 100 / list_size
+            stop = (index + 1) * 100 / list_size
+            # the first and last element should go until -+ infinity
+            if index == 0:
+                start = -inf
+            if index == list_size - 1:
+                stop = inf
+            speed_ranges[speed] = SpeedRange(start, target, stop)
+        self.speed_ranges = speed_ranges
+
+    def speed_to_homekit(self, speed):
+        """Map Home Assistant speed state to HomeKit speed."""
+        speed_range = self.speed_ranges.get(speed, None)
+        return speed_range.target if speed_range else 0
+
+    def speed_to_states(self, speed):
+        """Map HomeKit speed to Home Assistant speed state."""
+        for state, speed_range in self.speed_ranges.items():
+            if speed_range.start <= speed < speed_range.stop:
+                return state
+        # if nobody tampered with self.speed_ranges
+        # this happens only for +/-inf
+        raise KeyError(speed)
```

Comment:
With the removal of `inf`. This can probably be deleted as well.
Although this probably doesn't happen, with an `OrderedDict` we could return a default value
```py
return speed_ranges.keys()[0]

Reply:
Unfortunately, keys() is a view which cannot be indexed. However, the case still should not happen, so I keep raising a KeyError

Reply:
My fault, add `list(speed_ranges.keys())[0]`